# **Proyecto Integrador**
# ***LUMINA* - La Biblia Viva con IA**

Plataforma conversacional que facilita la comprensión y aplicación contextual del texto bíblico mediante NLP y RAG, con un enfoque responsable, trazable y centrado en el usuario.

---

> *“Porque Jehová da la sabiduría, y de su boca viene el conocimiento y la inteligencia.”*
>
> — Proverbios 2:6

---

### **Avance 2: Ingeniería de características (feature engineering)**
**Equipo:** *#19*  
**Curso:** *Proyecto Integrador*  
**Posgrado:** *MNA - Maestría en Inteligencia Artificial Aplicada*  
**Institución:** Tecnológico de Monterrey  
**Fecha:** *08 de febrero del 2026*  

---

### **Miembros del equipo**
- **A01732505** - Steven Sebastian Brutscher Cortez
- **A01795323** - Anghelo Daniel Pérez Martínez
- **A01423059** - Esmeralda González García

---

> *Este notebook corresponde al **Avance 2: Ingneiería de características** del curso Proyecto Integrador de la Maestría en Inteligencia Artificial Aplicada (MNA) del Tecnológico de Monterrey*

De acuerdo con **CRISP-ML(Q)**, este trabajo se ubica principalmente en la fase de:

- **Preparación de los Datos (Data Preparation)**, específicamente en la subfase de **Ingeniería de Características (Feature Engineering)**. En esta etapa, el objetivo es transformar un conjunto de datos ya validado y comprendido, resultado del Avance 1 (EDA), en un conjunto de características numéricas, estructurales y semánticas que permitan a los modelos de aprendizaje automático aprender de forma eficiente, justa y trazable.

- Esta fase constituye un puente crítico entre el análisis exploratorio y las etapas posteriores de modelado y evaluación, asegurando que las decisiones técnicas tomadas estén alineadas con los **objetivos del sistema** y con **principios de calidad, interpretabilidad y mitigación de sesgos**, tal como lo establece CRISP-ML(Q).

# **Introducción**

El presente entregable corresponde al **Avance 2 – Ingeniería de Características** del Proyecto Integrador *LUMINA – La Biblia Viva con IA*. En esta fase, se construye sobre el trabajo realizado en el Análisis Exploratorio de Datos (Avance 1), donde se validó la estructura canónica del corpus bíblico, se identificaron patrones, sesgos estructurales y complejidades semánticas, y se establecieron decisiones técnicas fundamentales para el diseño de un sistema de Recuperación Aumentada por Generación (RAG).

La ingeniería de características representa una de las etapas más determinantes en cualquier proyecto de aprendizaje automático, ya que define **cómo el conocimiento presente en los datos será representado, aprendido y recuperado por los modelos**. En el contexto de procesamiento de lenguaje natural (NLP), esta fase adquiere una relevancia aún mayor, dado que el texto debe ser transformado en representaciones numéricas capaces de **capturar significado, contexto y relaciones semánticas profundas**.

### **Objetivos de aprendizaje**

En concordancia con las instrucciones del curso y adaptados al contexto específico de *LUMINA*, los objetivos de este avance son:
- **Crear nuevas características** a partir del corpus bíblico que permitan representar de manera efectiva tanto su contenido semántico como su estructura canónica, con el fin de mejorar el desempeño y la estabilidad de los modelos posteriores.
- **Transformar y normalizar variables numéricas y categóricas relevantes**, garantizando un impacto equilibrado de las características auxiliares y facilitando la convergencia de los algoritmos de aprendizaje.
- **Mitigar riesgos de sesgo estructural**, previamente identificados en el EDA, mediante decisiones conscientes de representación, segmentación y codificación de los datos.
- **Aplicar técnicas de selección y extracción de características**, alineadas con las particularidades del NLP, para reducir complejidad, redundancia y requerimientos computacionales sin comprometer la fidelidad semántica del texto.
- **Fundamentar cada decisión técnica desde una perspectiva metodológica, ética y práctica**, asegurando trazabilidad y coherencia con la metodología CRISP-ML(Q).

### **Relación con el Avance 1 – Análisis Exploratorio de Datos**

El Avance 1 permitió comprender en profundidad la naturaleza del corpus bíblico utilizado en *LUMINA*, revelando aspectos clave como:
- La **fuerte desigualdad estructural entre libros**.
- La **alta complejidad léxica y semántica del texto**.
- La **presencia de valores atípicos textuales con alta densidad teológica**.
- Los **riesgos inherentes de sesgo en procesos de recuperación ingenuos**.

Este segundo avance no replantea dichos hallazgos, sino que los utiliza como base para diseñar una ingeniería de características informada, donde cada transformación aplicada responde directamente a un problema o necesidad identificada durante el EDA.

### **Conexión con CRISP-ML(Q)**

Dentro de CRISP-ML(Q), la "ingeniería de características" forma parte esencial de la fase de Preparación de los Datos, y su correcta ejecución impacta directamente en:
- La **calidad del modelado**.
- La **interpretabilidad de los resultados**.
- La **robustez del sistema frente a sesgos**.
- La **viabilidad de su despliegue futuro**.

En este sentido, el presente notebook busca no solo cumplir con los requisitos técnicos del curso, sino también sentar bases sólidas para un sistema de IA responsable, trazable y alineado con principios de calidad a lo largo de todo su ciclo de vida.

---

> *“Escudriñad las Escrituras; porque a vosotros os parece que en ellas tenéis la vida eterna; y ellas son las que dan testimonio de mí.”*
>
> — Juan 5:39

---

# **Parte 1 – Contexto del dataset**

Antes de aplicar cualquier técnica de ingeniería de características, es necesario establecer con claridad el **contexto del conjunto de datos** que será transformado. Esta sección resume los elementos esenciales del corpus bíblico utilizado, enfatizando aquellos aspectos que influyen directamente en las decisiones de representación y transformación que se abordarán en las siguientes partes.

### **Descripción general del corpus**

El dataset utilizado corresponde a una versión estructurada de la Biblia en formato tabular, donde **cada fila representa un versículo único**, identificado de manera canónica por su libro, capítulo y versículo. El corpus contiene un total de 30,938 versículos, abarcando los 66 libros de la Biblia, distribuidos entre el Antiguo y el Nuevo Testamento.

Cada versículo cuenta con dos representaciones textuales, pertenecientes a dos traducciones en español:
- **Reina-Valera 1909 (RV1909)**
- **Biblia en Lenguaje Actual / Versión en lenguaje sencillo (VBL)**

Esta coexistencia de traducciones permite explorar diferencias estilísticas y semánticas, pero también introduce retos adicionales para la representación y el retrieval.

### **Estructura y variables relevantes**

Las principales variables disponibles en el dataset incluyen:
- Identificadores canónicos (libro, capítulo, versículo, referencia textual).
- Variables categóricas estructurales (libro, testamento).
- Variables textuales (versículos en distintas traducciones).
- Variables numéricas derivables (longitud del texto, conteo de palabras, etc.).

Desde la perspectiva de la ingeniería de características, estas variables pueden agruparse en dos grandes categorías:
1. **Variables semánticas**, cuyo contenido principal es el texto bíblico.
2. **Variables estructurales y auxiliares**, que aportan contexto, organización y señales adicionales para el aprendizaje automático.

### **Implicaciones para la ingeniería de características**

El análisis exploratorio previo mostró que el corpus presenta:

- Alta dimensionalidad semántica.
- Distribuciones de longitud con colas largas.
- Sesgos estructurales por libro y testamento.

Estas características hacen inviable el uso de representaciones textuales simples (como bag-of-words) para un sistema RAG, y justifican el uso de **técnicas de extracción semántica avanzadas**, así como la incorporación consciente de **metadatos estructurales** como características complementarias.

Por tanto, la ingeniería de características en LUMINA no se limita a transformar texto en números, sino que busca equilibrar significado, estructura y fidelidad textual, preparando el dataset para etapas posteriores de modelado y recuperación semántica.

---

> *“Entiende primero lo que has de hacer, y después edifica tu casa.”*
>
> — Proverbios 24:27

---

# **Parte 2 – Generación de características (Feature Engineering)**

In [ ]:
# ============================================================
# Montando Drive
# ============================================================

from google.colab import drive
drive.mount('/content/drive')

In [ ]:
# ============================================================
# Inicializando las librerías necesarias
# ============================================================

import os
import re
from datetime import datetime

import numpy as np
import pandas as pd

from tqdm.auto import tqdm

import json

!pip -q install sentence-transformers
from sentence_transformers import SentenceTransformer

### **2.1 Carga del dataset y preprocesamiento mínimo**

En esta sección se realiza la **carga del corpus estructurado** y un **preprocesamiento mínimo y consciente**, alineado con el enfoque de *LUMINA*: preservar la fidelidad textual y la trazabilidad del contenido bíblico.

A diferencia de pipelines típicos de NLP, aquí **no se aplica lematización, stemming ni normalización agresiva**, ya que dichas transformaciones podrían alterar matices semánticos importantes para el objetivo del sistema RAG. En su lugar, se aplican únicamente operaciones seguras y reversibles, como:

- Normalización de espacios en blanco (colapsar espacios múltiples).
- Conversión controlada de valores nulos a cadenas vacías.
- Verificaciones básicas de integridad en columnas textuales.

Además, se definen las rutas actualizadas para acceder al dataset en Google Drive y para guardar los artefactos generados en el avance anterior (EDA).

In [ ]:
# ============================================================
# Parte 2.1 — Carga del dataset y preprocesamiento mínimo
# ============================================================

# -----------------------------
# 0) Rutas (actualizadas)
# -----------------------------

# CSV procesado (ubicación general para uso en avances futuros)
DATA_PATH = "/content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/data/processed/bible_structured.csv"

# Carpeta del Avance 2 (ruta para guardar artefactos generados en este notebook)
OUTPUT_DIR = "/content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/Avance 2 - Ingeniería de características (feature engineering)"

# Subcarpetas para mantener orden
ARTIFACTS_DIR = os.path.join(OUTPUT_DIR, "artifacts")
EMB_DIR = os.path.join(OUTPUT_DIR, "embeddings")

os.makedirs(ARTIFACTS_DIR, exist_ok=True)
os.makedirs(EMB_DIR, exist_ok=True)

print("DATA_PATH:", DATA_PATH)
print("OUTPUT_DIR:", OUTPUT_DIR)
print("ARTIFACTS_DIR:", ARTIFACTS_DIR)
print("EMB_DIR:", EMB_DIR)

# -----------------------------
# 1) Carga del dataset
# -----------------------------

df = pd.read_csv(DATA_PATH)

print("Dataset cargado correctamente")
print("Shape:", df.shape)
print("Columnas:", df.columns.tolist())

df.head(3)

# -----------------------------
# 2) Preprocesamiento mínimo
# -----------------------------
# Nota metodológica:
# - No lematizamos, no hacemos stemming, no quitamos stopwords.
# - Solo normalizamos espacios y controlamos nulos para consistencia.

def normalize_whitespace(text: str) -> str:
    """Normalización suave: colapsa espacios múltiples y recorta extremos."""
    if pd.isna(text):
        return ""
    text = str(text)
    text = re.sub(r"\s+", " ", text).strip()
    return text

# Aplicamos normalización de espacios a ambas traducciones (sin alterar signos, tildes ni mayúsculas)
df["text_rv1909"] = df["text_rv1909"].map(normalize_whitespace)
df["text_vbl"] = df["text_vbl"].map(normalize_whitespace)

# Verificaciones básicas de integridad
print("\nVerificación rápida de nulos / vacíos")
print("Nulos RV1909:", df["text_rv1909"].isna().sum())
print("Nulos VBL:", df["text_vbl"].isna().sum())

print("\nVacíos RV1909 ('' tras normalización):", (df["text_rv1909"] == "").sum())
print("Vacíos VBL ('' tras normalización):", (df["text_vbl"] == "").sum())

# Muestra rápida para confirmar que no se alteró el contenido (solo espacios)
df.sample(3, random_state=42)[["ref", "book_es", "chapter", "verse", "text_rv1909", "text_vbl"]]

DATA_PATH: /content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/data/processed/bible_structured.csv
OUTPUT_DIR: /content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/Avance 2 - Ingeniería de características (feature engineering)
ARTIFACTS_DIR: /content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/Avance 2 - Ingeniería de características (feature engineering)/artifacts
EMB_DIR: /content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/Avance 2 - Ingeniería de características (feature engineering)/embeddings
Dataset cargado correctamente
Shape: (30938, 7)
Columnas: ['ref', 'book', 'chapter', 'verse', 'text_rv1909', 'text_vbl', 'book_es']

Verificación rápida de nulos / vacíos
Nulos RV1909: 0
Nulos VBL: 0

Vacíos RV1909 ('' tras normalización): 0
Vacíos VBL ('' tras normalización): 0


,ref,book_es,chapter,verse,text_rv1909,text_vbl
12738,EST 7:8,Ester,7,8,Volvió después el rey del huerto del palacio a...,Cuando el rey regresó del jardín del palacio a...
12491,NEH 10:19,Nehemías,10,19,"Ariph, Anathoth, Nebai,","Harif, Anatot, Nebai,"
22150,HOS 12:9,Oseas,12,9,"Y dijo Ephraim: Ciertamente yo he enriquecido,...","Pero yo soy el Señor tu Dios, que te saqué de ..."


#### **Interpretación del resultado (2.1)**

El resultado confirma que el dataset está listo para ingeniería de características, sin introducir ruido ni alteraciones semánticas indebidas.

- El dataset contiene 30,938 filas.
- Las 7 columnas esperadas están presentes y correctamente nombradas.
- Tanto text_rv1909 como text_vbl presentan 0 valores nulos.

Tras la normalización de espacios, no se generaron textos vacíos, lo que confirma que el preprocesamiento aplicado fue seguro y no destructivo. Las muestras visualizadas muestran texto original intacto, diferencias estilísticas coherentes entre RV1909 y VBL, y ninguna pérdida de puntuación, mayúsculas o contenido semántico.

La normalización aplicada fue estrictamente estructural (espacios), no lingüística. Esto respeta el principio fundamental de *LUMINA*: el texto bíblico no se “corrige”, se representa.

Tenemos un corpus estable, confiable y preparado para la creación de nuevas características.

---

> *“Bueno es adquirir sabiduría, y mejor es adquirir inteligencia.”*
>
> — Proverbios 16:16

---

### **2.2 Features estructurales (metadatos canónicos) y numéricas auxiliares**

En esta sección se construyen **características estructurales y numéricas auxiliares** que complementan la representación semántica del texto bíblico. Estas características no buscan sustituir los embeddings, sino **aportar contexto canónico y señales cuantitativas que permitan**:

- Mitigar sesgos estructurales identificados en el EDA.
- Facilitar reglas posteriores de segmentación (chunking).
- Mejorar la interpretabilidad y trazabilidad del sistema.

Se incluyen:

- **Metadatos canónicos derivados**, como testamento y tipo de libro.
- **Variables numéricas auxiliares**, como longitud del versículo y diferencias entre traducciones.
- **Discretización (binning)** de longitudes, aplicada únicamente a variables numéricas, nunca al texto crudo.

Todas las decisiones están fundamentadas en los hallazgos del Análisis Exploratorio de Datos.

In [ ]:
# ============================================================
# Parte 2.2 — Features estructurales y numéricas auxiliares
# ============================================================

# -----------------------------
# 1) Features estructurales: Testamento
# -----------------------------
# Inferimos el testamento a partir del orden canónico detectado en el dataset.
# (39 libros AT, 27 NT)

books_in_order = (
    df[["book", "book_es"]]
    .drop_duplicates()
    .reset_index(drop=True)
)

if len(books_in_order) != 66:
    print("Advertencia: se esperaban 66 libros, se detectaron:", len(books_in_order))

books_in_order["testament"] = (
    ["AT"] * min(39, len(books_in_order)) +
    ["NT"] * max(0, len(books_in_order) - 39)
)

book_to_testament = dict(zip(books_in_order["book"], books_in_order["testament"]))
df["testament"] = df["book"].map(book_to_testament)

print("Distribución por testamento:")
print(df["testament"].value_counts())

# -----------------------------
# 2) Features estructurales: Clasificación canónica
# -----------------------------
# Esta taxonomía sigue una clasificación bíblica estándar y es útil para:
# - chunking adaptativo más fino
# - análisis de sesgo por grupos canónicos
# - control del retrieval (metadatos)

CANON_GROUP_MAP_ES = {
    # Antiguo Testamento — Pentateuco (Ley)
    "Génesis": "pentateuco",
    "Éxodo": "pentateuco",
    "Levítico": "pentateuco",
    "Números": "pentateuco",
    "Deuteronomio": "pentateuco",

    # Antiguo Testamento — Históricos
    "Josué": "historicos",
    "Jueces": "historicos",
    "Rut": "historicos",
    "1 Samuel": "historicos",
    "2 Samuel": "historicos",
    "1 Reyes": "historicos",
    "2 Reyes": "historicos",
    "1 Crónicas": "historicos",
    "2 Crónicas": "historicos",
    "Esdras": "historicos",
    "Nehemías": "historicos",
    "Ester": "historicos",

    # Antiguo Testamento — Poéticos / Sabiduría
    "Job": "poeticos_sabiduria",
    "Salmos": "poeticos_sabiduria",
    "Proverbios": "poeticos_sabiduria",
    "Eclesiastés": "poeticos_sabiduria",
    "Cantares": "poeticos_sabiduria",
    "Cantar de los Cantares": "poeticos_sabiduria",

    # Antiguo Testamento — Profetas mayores
    "Isaías": "profetas_mayores",
    "Jeremías": "profetas_mayores",
    "Lamentaciones": "profetas_mayores",
    "Ezequiel": "profetas_mayores",
    "Daniel": "profetas_mayores",

    # Antiguo Testamento — Profetas menores
    "Oseas": "profetas_menores",
    "Joel": "profetas_menores",
    "Amós": "profetas_menores",
    "Abdías": "profetas_menores",
    "Jonás": "profetas_menores",
    "Miqueas": "profetas_menores",
    "Nahúm": "profetas_menores",
    "Habacuc": "profetas_menores",
    "Sofonías": "profetas_menores",
    "Hageo": "profetas_menores",
    "Zacarías": "profetas_menores",
    "Malaquías": "profetas_menores",

    # Nuevo Testamento — Evangelios
    "Mateo": "evangelios",
    "Marcos": "evangelios",
    "Lucas": "evangelios",
    "Juan": "evangelios",

    # Nuevo Testamento — Libro histórico (Hechos)
    "Hechos de los Apóstoles": "hechos",

    # Nuevo Testamento — Cartas paulinas (13)
    "Romanos": "cartas_paulinas",
    "1 Corintios": "cartas_paulinas",
    "2 Corintios": "cartas_paulinas",
    "Gálatas": "cartas_paulinas",
    "Efesios": "cartas_paulinas",
    "Filipenses": "cartas_paulinas",
    "Colosenses": "cartas_paulinas",
    "1 Tesalonicenses": "cartas_paulinas",
    "2 Tesalonicenses": "cartas_paulinas",
    "1 Timoteo": "cartas_paulinas",
    "2 Timoteo": "cartas_paulinas",
    "Tito": "cartas_paulinas",
    "Filemón": "cartas_paulinas",

    # Nuevo Testamento — Cartas generales (8) + Hebreos (la tratamos como general para evitar debate de autoría)
    "Hebreos": "cartas_generales",
    "Santiago": "cartas_generales",
    "1 Pedro": "cartas_generales",
    "2 Pedro": "cartas_generales",
    "1 Juan": "cartas_generales",
    "2 Juan": "cartas_generales",
    "3 Juan": "cartas_generales",
    "Judas": "cartas_generales",

    # Nuevo Testamento — Apocalipsis
    "Apocalipsis": "apocalipsis",
}

df["canon_group"] = df["book_es"].map(CANON_GROUP_MAP_ES).fillna("otro")

print("\nDistribución por canon_group:")
print(df["canon_group"].value_counts())

# Control de calidad: libros no mapeados:
unmapped = sorted(df.loc[df["canon_group"] == "otro", "book_es"].unique().tolist())
print("\nLibros no mapeados explícitamente:", unmapped)

# Versión "coarse" para análisis más general
CANON_GROUP_COARSE = {
    "pentateuco": "AT_ley",
    "historicos": "AT_historia",
    "poeticos_sabiduria": "AT_poesia_sabiduria",
    "profetas_mayores": "AT_profetas",
    "profetas_menores": "AT_profetas",
    "evangelios": "NT_evangelios",
    "hechos": "NT_historia",
    "cartas_paulinas": "NT_cartas",
    "cartas_generales": "NT_cartas",
    "apocalipsis": "NT_apocaliptico",
    "otro": "otro"
}

df["canon_group_coarse"] = df["canon_group"].map(CANON_GROUP_COARSE).fillna("otro")

# -----------------------------
# 3) Features numéricas auxiliares (longitudes)
# -----------------------------
WORD_RE = re.compile(r"\b\w+\b", flags=re.UNICODE)

def word_count(text: str) -> int:
    if not text:
        return 0
    return len(WORD_RE.findall(text))

df["wc_rv1909"] = df["text_rv1909"].map(word_count)
df["wc_vbl"] = df["text_vbl"].map(word_count)

# Diferencia de longitudes entre traducciones
df["wc_diff_vbl_minus_rv"] = df["wc_vbl"] - df["wc_rv1909"]

# -----------------------------
# 4) Discretización (binning) de longitudes
# -----------------------------
bins = [-1, 8, 16, 32, 64, 128, 10_000]
labels = ["muy_corto", "corto", "medio", "largo", "muy_largo", "extremo"]

df["len_bin_rv1909"] = pd.cut(df["wc_rv1909"], bins=bins, labels=labels)
df["len_bin_vbl"] = pd.cut(df["wc_vbl"], bins=bins, labels=labels)

df[[
    "ref", "book_es", "testament", "canon_group", "canon_group_coarse",
    "wc_rv1909", "wc_vbl", "wc_diff_vbl_minus_rv",
    "len_bin_rv1909", "len_bin_vbl"
]].sample(5, random_state=42)

Distribución por testamento:
testament
AT    22999
NT     7939
Name: count, dtype: int64

Distribución por canon_group:
canon_group
historicos            6974
pentateuco            5819
poeticos_sabiduria    4766
profetas_mayores      4428
evangelios            3768
cartas_paulinas       2031
profetas_menores      1012
hechos                1002
cartas_generales       734
apocalipsis            404
Name: count, dtype: int64

Libros no mapeados explícitamente: []


,ref,book_es,testament,canon_group,canon_group_coarse,wc_rv1909,wc_vbl,wc_diff_vbl_minus_rv,len_bin_rv1909,len_bin_vbl
12738,EST 7:8,Ester,AT,historicos,AT_historia,53,53,0,largo,largo
12491,NEH 10:19,Nehemías,AT,historicos,AT_historia,3,3,0,muy_corto,muy_corto
22150,HOS 12:9,Oseas,AT,profetas_menores,AT_profetas,23,30,7,medio,medio
11453,2CH 16:12,2 Crónicas,AT,historicos,AT_historia,28,38,10,medio,largo
2346,EXO 29:16,Éxodo,AT,pentateuco,AT_ley,14,10,-4,corto,corto


#### **Interpretación del resultado (2.2)**

En esta sección se construyeron características estructurales y numéricas auxiliares que enriquecen la representación del corpus bíblico sin alterar su contenido semántico. Se derivaron metadatos canónicos como el testamento y una clasificación jerárquica de los libros bíblicos, logrando una cobertura completa del corpus sin valores no mapeados.

Asimismo, se generaron variables numéricas relacionadas con la longitud de los versículos y las diferencias entre traducciones, junto con una discretización controlada de dichas longitudes. Estas características permiten mitigar sesgos estructurales, facilitan la segmentación adaptativa del texto y preparan el dataset para procesos posteriores de retrieval y modelado semántico.

---

> *“El corazón del entendido adquiere sabiduría, y el oído de los sabios busca la ciencia.”*
>
> — Proverbios 18:15

---

### **2.3 Chunking adaptativo (pasajes) — Feature Engineering para RAG**

En esta sección se implementa un **chunking adaptativo**, cuyo objetivo es transformar versículos individuales en **unidades semánticas recuperables** (pasajes), adecuadas para un sistema de Recuperación Aumentada por Generación (RAG).

A partir de los hallazgos del EDA y de la clasificación canónica construida en la sección anterior, se define una estrategia de segmentación que **no es global ni uniforme**, sino que se ajusta al tipo de contenido bíblico. Esta decisión responde a que distintos géneros presentan densidades semánticas y estructuras discursivas diferentes.

El chunking adaptativo busca:
- Preservar coherencia semántica.
- Evitar fragmentación excesiva en textos densos.
- Reducir el sesgo de recuperación hacia libros largos.
- Preparar el texto para la generación de embeddings y búsqueda semántica.

In [ ]:
# ============================================================
# Parte 2.3 — Chunking adaptativo (pasajes) para RAG
# ============================================================

# -----------------------------
# 1) Reglas de chunking por grupo canónico
# -----------------------------
# Valores expresados como número de versículos por chunk.
# Se elige un valor medio dentro de un rango razonable por grupo.

CHUNK_SIZE_BY_CANON_GROUP = {
    "pentateuco": 4,
    "historicos": 4,
    "poeticos_sabiduria": 2,
    "profetas_mayores": 4,
    "profetas_menores": 3,
    "evangelios": 3,
    "hechos": 4,
    "cartas_paulinas": 3,
    "cartas_generales": 3,
    "apocalipsis": 2,
    "otro": 3,
}

def get_chunk_size(canon_group: str) -> int:
    """Devuelve el tamaño de chunk según el grupo canónico."""
    return CHUNK_SIZE_BY_CANON_GROUP.get(canon_group, 3)

# -----------------------------
# 2) Preparación del dataframe
# -----------------------------
# Orden canónico + orden interno
df_sorted = df.sort_values(
    ["book", "chapter", "verse"]
).reset_index(drop=True)

# -----------------------------
# 3) Construcción de chunks
# -----------------------------
chunks = []
chunk_id = 0

for (book, chapter), g in tqdm(
    df_sorted.groupby(["book", "chapter"], sort=False),
    desc="Construyendo chunks adaptativos"
):
    g = g.sort_values("verse").reset_index(drop=True)

    canon_group = g["canon_group"].mode().iloc[0]
    testament = g["testament"].mode().iloc[0]
    book_es = g["book_es"].mode().iloc[0]
    chunk_size = get_chunk_size(canon_group)

    for start in range(0, len(g), chunk_size):
        block = g.iloc[start:start + chunk_size]

        verse_start = int(block["verse"].iloc[0])
        verse_end = int(block["verse"].iloc[-1])

        if verse_start == verse_end:
            chunk_ref = f"{book} {chapter}:{verse_start}"
        else:
            chunk_ref = f"{book} {chapter}:{verse_start}-{verse_end}"

        chunks.append({
            "chunk_id": f"chk_{chunk_id:06d}",
            "book": book,
            "book_es": book_es,
            "testament": testament,
            "canon_group": canon_group,
            "canon_group_coarse": block["canon_group_coarse"].mode().iloc[0],
            "chapter": int(chapter),
            "verse_start": verse_start,
            "verse_end": verse_end,
            "chunk_ref": chunk_ref,
            "chunk_text_rv1909": " ".join(block["text_rv1909"].tolist()),
            "chunk_text_vbl": " ".join(block["text_vbl"].tolist()),
        })

        chunk_id += 1

df_chunks = pd.DataFrame(chunks)

print("Número total de chunks creados:", df_chunks.shape[0])
df_chunks.sample(5, random_state=42)

Construyendo chunks adaptativos:   0%|          | 0/1188 [00:00<?, ?it/s]

Número total de chunks creados: 10058


,chunk_id,book,book_es,testament,canon_group,canon_group_coarse,chapter,verse_start,verse_end,chunk_ref,chunk_text_rv1909,chunk_text_vbl
8292,chk_008292,PSA,Salmos,AT,poeticos_sabiduria,AT_poesia_sabiduria,9,13,14,PSA 9:13-14,"Ten misericordia de mí, Jehová: mira mi aflicc...","¡Ten misericordia de mi, oh, Señor! ¡Mira cómo..."
3127,chk_003127,EZK,Ezequiel,AT,profetas_mayores,AT_profetas,29,9,12,EZK 29:9-12,Y la tierra de Egipto será asolada y desierta:...,Egipto se convertirá en un páramo vacío. Enton...
8507,chk_008507,PSA,Salmos,AT,poeticos_sabiduria,AT_poesia_sabiduria,37,3,4,PSA 37:3-4,"Espera en Jehová, y haz bien; vivirás en la ti...","Confía en el Señor, y haz el bien. Vive en la ..."
2178,chk_002178,DEU,Deuteronomio,AT,pentateuco,AT_ley,3,5,8,DEU 3:5-8,Todas éstas eran ciudades fortalecidas con alt...,Todas estas ciudades estaban fortificadas con ...
107,chk_000107,1CH,1 Crónicas,AT,historicos,AT_historia,11,13,16,1CH 11:13-16,"Este estuvo con David en Pasdammin, estando al...",Estaba con David en Pasdamin cuando los filist...


#### **Interpretación del resultado (2.3)**

En esta sección se implementó una estrategia de chunking adaptativo, transformando el corpus bíblico en unidades semánticas coherentes (pasajes) adecuadas para un sistema de Recuperación Aumentada por Generación (RAG).

La segmentación se basó en la clasificación canónica previamente construida, permitiendo ajustar el tamaño de los chunks según la densidad semántica de cada grupo bíblico. Como resultado, se generaron 10,058 chunks, cada uno con contexto suficiente, trazabilidad canónica y continuidad textual.

Este proceso reduce el sesgo estructural, mejora la recuperación semántica y prepara el corpus para la extracción de características densas mediante embeddings.

---

> *“Las palabras del sabio son como aguijones, y como clavos hincados las de los maestros de las congregaciones.”*
>
> — Eclesiastés 12:11

---

### **2.4 Codificación de variables categóricas y extracción de características semánticas (Embeddings)**

En esta sección se realiza la extracción de características semánticas densas (embeddings) a partir de los chunks construidos previamente.

Mientras que las variables estructurales aportan contexto y control, los embeddings permiten representar el significado profundo del texto en un espacio vectorial continuo, facilitando la comparación semántica entre pasajes bíblicos y consultas del usuario.

Asimismo, se codifican variables categóricas relevantes para su posible uso en modelos híbridos o análisis posteriores. Los embeddings se generan de manera separada para cada traducción, preservando diferencias estilísticas y semánticas entre versiones.

In [ ]:
# ============================================================
# Parte 2.4 — Codificación categórica y embeddings semánticos
# ============================================================

# -----------------------------
# 1) Codificación de variables categóricas (One-Hot)
# -----------------------------
# Nota: no se aplican a los embeddings directamente, sino como features auxiliares

categorical_cols = ["testament", "canon_group", "canon_group_coarse"]
df_chunks_encoded = pd.get_dummies(
    df_chunks,
    columns=categorical_cols,
    prefix=categorical_cols
)

print("Shape con variables categóricas codificadas:", df_chunks_encoded.shape)

# Guardamos versión tabular (sin embeddings)
timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
tabular_path = f"{ARTIFACTS_DIR}/lumina_chunks_tabular_{timestamp}.parquet"
df_chunks_encoded.to_parquet(tabular_path, index=False)

print("Dataset tabular guardado en:", tabular_path)

# -----------------------------
# 2) Extracción de embeddings semánticos
# -----------------------------
# Usamos un modelo multilingüe robusto, adecuado para español y retrieval semántico

MODEL_NAME = "sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2"
model = SentenceTransformer(MODEL_NAME)

def embed_texts(texts, batch_size=64):
    return model.encode(
        texts,
        batch_size=batch_size,
        show_progress_bar=True,
        convert_to_numpy=True,
        normalize_embeddings=True  # útil para similitud coseno
    )

# Textos por traducción
texts_rv = df_chunks["chunk_text_rv1909"].tolist()
texts_vbl = df_chunks["chunk_text_vbl"].tolist()

# Generación de embeddings
emb_rv = embed_texts(texts_rv)
emb_vbl = embed_texts(texts_vbl)

print("Embeddings RV1909:", emb_rv.shape)
print("Embeddings VBL:", emb_vbl.shape)

# -----------------------------
# 3) Guardado de embeddings
# -----------------------------
embeddings_path = f"{EMB_DIR}/lumina_embeddings_{timestamp}.npz"
np.savez_compressed(
    embeddings_path,
    chunk_id=df_chunks["chunk_id"].to_numpy(),
    emb_rv1909=emb_rv,
    emb_vbl=emb_vbl
)

print("Embeddings guardados en:", embeddings_path)

# Manifest de trazabilidad
manifest = {
    "model": MODEL_NAME,
    "normalized": True,
    "n_chunks": int(df_chunks.shape[0]),
    "created_at": timestamp
}

manifest_path = f"{EMB_DIR}/embedding_manifest_{timestamp}.json"
with open(manifest_path, "w", encoding="utf-8") as f:
    json.dump(manifest, f, ensure_ascii=False, indent=2)

print("Manifest guardado en:", manifest_path)

Shape con variables categóricas codificadas: (10058, 29)
Dataset tabular guardado en: /content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/Avance 2 - Ingeniería de características (feature engineering)/artifacts/lumina_chunks_tabular_20260207_084138.parquet


/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/229 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

README.md: 0.00B [00:00, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/645 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/471M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/199 [00:00<?, ?it/s]

BertModel LOAD REPORT from: sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2
Key                     | Status     |  | 
------------------------+------------+--+-
embeddings.position_ids | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


tokenizer_config.json:   0%|          | 0.00/526 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/9.08M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/239 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Batches:   0%|          | 0/158 [00:00<?, ?it/s]

Batches:   0%|          | 0/158 [00:00<?, ?it/s]

Embeddings RV1909: (10058, 384)
Embeddings VBL: (10058, 384)
Embeddings guardados en: /content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/Avance 2 - Ingeniería de características (feature engineering)/embeddings/lumina_embeddings_20260207_084138.npz
Manifest guardado en: /content/drive/MyDrive/Maestría ITESM/LUMINA - Proyecto Integrador/Avance 2 - Ingeniería de características (feature engineering)/embeddings/embedding_manifest_20260207_084138.json


#### **Conclusión final de la Parte 2 – Generación de características**

En esta parte se llevó a cabo la generación completa de características para el corpus bíblico de LUMINA, transformando el texto original en representaciones estructuradas y semánticas adecuadas para aprendizaje automático y recuperación de información.

Se construyeron unidades semánticas coherentes (chunks) mediante una estrategia de segmentación adaptativa basada en la clasificación canónica del texto, logrando un equilibrio entre contexto, granularidad y trazabilidad. Asimismo, se codificaron variables categóricas relevantes y se extrajeron embeddings semánticos densos para cada pasaje, utilizando un modelo multilingüe robusto y eficiente.

Como resultado, se generaron artefactos persistentes y reproducibles que representan el conocimiento bíblico tanto en forma tabular como vectorial, dejando el dataset completamente preparado para tareas posteriores de retrieval semántico y modelado dentro de un sistema RAG.

---

> *“Considera lo que digo, y el Señor te dé entendimiento en todo.”*
>
> — 2 Timoteo 2:7

---

# **Parte 3 – Normalización y transformación**

En esta sección se analizan las decisiones relacionadas con la **normalización, escalamiento y transformación de características**, en el contexto de un proyecto de procesamiento de lenguaje natural basado en embeddings.

A diferencia de datasets puramente numéricos, en LUMINA coexisten distintos tipos de características: variables semánticas densas (embeddings), variables numéricas auxiliares y variables categóricas codificadas. Por ello, no todas las transformaciones comunes son aplicables ni deseables.

El objetivo de esta parte es **justificar de forma explícita**:
- qué características se normalizan o escalan,
- cuáles se dejan sin transformar,
- y por qué estas decisiones favorecen la estabilidad, equidad y convergencia de modelos posteriores.

#### **¿Qué se escala y qué no?**

**No se escalan:**

**- Embeddings semánticos**
  - Ya están normalizados ('normalize_embeddings'=True)
  - Escalarlos rompería relaciones angulares (coseno)

**Sí se escalan** (si se usan en modelos híbridos):

- Longitudes ('wc_rv1909', 'wc_vbl')
- Diferencias de longitud

En este avance **no se aplica escalamiento explícito**, porque:

- no se entrenan modelos aún,
- se prioriza preservar interpretabilidad,
- el escalamiento se ejecuta típicamente *justo antes del entrenamiento*.

#### **Transformaciones (log, Box-Cox, etc.)**

No se aplican transformaciones no lineales al texto ni a los embeddings, ya que:

- el texto ya fue proyectado a un espacio latente semántico,
- las transformaciones podrían distorsionar similitudes semánticas,
- y no aportan beneficios claros en esta fase.

Las transformaciones quedan **fuera del alcance del Avance 2** y se considerarían únicamente en contextos de modelado supervisado posterior.

---

> *“Examinadlo todo; retened lo bueno.”*
>
> — 1 Tesalonicenses 5:21

---

# **Parte 4 – Selección y extracción de características**

En esta sección se abordan las decisiones relacionadas con la selección y extracción de características, con el objetivo de reducir complejidad, mitigar sesgos y asegurar que el conjunto final de variables sea adecuado para tareas posteriores de recuperación semántica y modelado.

En el contexto de procesamiento de lenguaje natural moderno, la selección de características difiere sustancialmente de enfoques clásicos basados en variables tabulares independientes. En *LUMINA*, el núcleo informativo del sistema reside en representaciones semánticas densas (embeddings), mientras que las características estructurales y numéricas cumplen un rol complementario de control, balance y trazabilidad.

Por ello, esta parte se centra tanto en justificar por qué ciertos métodos clásicos no son apropiados, como en explicar por qué los embeddings constituyen una forma legítima y poderosa de extracción de características.

#### **4.1 Selección de características: por qué NO aplicar filtros estadísticos clásicos**

**Explicación conceptual**

En problemas tradicionales de machine learning supervisado, es común aplicar técnicas de selección de características como:

- Umbral de varianza
- Correlación entre variables
- Pruebas Chi-cuadrado
- ANOVA

Sin embargo, estas técnicas **no son adecuadas** para el núcleo semántico de *LUMINA* por las siguientes razones:

1. **Los embeddings no representan variables independientes**   
Cada dimensión de un embedding no tiene un significado semántico aislado; el significado emerge de la *relación conjunta entre dimensiones*.

2. **No existe una variable objetivo en esta fase**   
El proyecto se encuentra aún en la fase de Data Preparation dentro de CRISP-ML(Q). No hay etiquetas ni tareas supervisadas que permitan evaluar relevancia estadística clásica.

3. **Aplicar filtros sobre embeddings rompería la geometría semántica**   
Eliminar dimensiones por varianza o correlación puede distorsionar las distancias angulares, afectando negativamente la similitud coseno utilizada en retrieval.

Por estas razones, **no se aplican técnicas de selección de características clásicas sobre los embeddings**, y esta omisión es una **decisión metodológica consciente**, no una limitación técnica.

**¿Qué sí se selecciona explícitamente?**

La selección de características en *LUMINA* se realiza **a nivel de tipos de variables**, no de dimensiones individuales:

- **Se conservan:**
  - Embeddings semánticos completos (RV1909 y VBL).
  - Metadatos canónicos ('testament', 'canon_group', etc.).

- **Se consideran auxiliares:**
  - Longitudes y bins, útiles solo en modelos híbridos o análisis posteriores.

- **Se excluyen del modelado semántico directo:**
  - Variables puramente descriptivas que no aportan significado.

#### **4.2 Extracción de características: embeddings como representación latente**

**Explicación conceptual**

La extracción de características consiste en transformar los datos originales en una representación más compacta y significativa. En *LUMINA*, esta transformación se realizó mediante **embeddings semánticos**, que proyectan cada chunk bíblico a un espacio vectorial continuo de baja dimensionalidad.

Desde una perspectiva metodológica, los embeddings cumplen el mismo rol que técnicas clásicas de extracción como PCA o análisis factorial, pero con ventajas clave:

- Capturan **significado contextual**, no solo varianza.
- Preservan relaciones semánticas no lineales.
- Son independientes del idioma superficial del texto.

Por tanto, el uso de embeddings constituye una **técnica de extracción de características avanzada**, alineada con el estado del arte en NLP.

#### **4.3 PCA exploratorio sobre embeddings (análisis, no reducción)**

**Justificación**

Aunque **no se utiliza PCA para reducir dimensionalidad en producción**, puede aplicarse de forma exploratoria para:

- Verificar que los embeddings no colapsan.
- Observar separación semántica general entre grupos canónicos.
- Detectar anomalías evidentes.

Este análisis **no modifica el dataset**, solo sirve como validación cualitativa.

In [ ]:
# ============================================================
# Parte 4.3 — PCA exploratorio sobre embeddings (análisis)
# ============================================================

from sklearn.decomposition import PCA

# Usamos una submuestra para evitar costos computacionales altos
sample_size = 2000
rng = np.random.default_rng(42)

idx = rng.choice(len(emb_rv), size=sample_size, replace=False)
emb_sample = emb_rv[idx]

pca = PCA(n_components=2, random_state=42)
emb_2d = pca.fit_transform(emb_sample)

print("Varianza explicada por los 2 primeros componentes:")
print(pca.explained_variance_ratio_)
print("Varianza acumulada:", pca.explained_variance_ratio_.sum())

Varianza explicada por los 2 primeros componentes:
[0.09890959 0.0796361 ]
Varianza acumulada: 0.17854568


#### **Interpretación**

La varianza explicada por los dos primeros componentes principales (~17.9 %) es consistente con lo esperado en representaciones semánticas densas. En embeddings de lenguaje natural, el significado se distribuye a lo largo de múltiples dimensiones, por lo que una varianza acumulada moderada en las primeras componentes indica diversidad semántica y ausencia de colapso representacional.

Estos resultados sugieren que los embeddings capturan información rica y no lineal, confirmando que el espacio vectorial es adecuado para tareas de similitud semántica y recuperación de información, sin justificar una reducción de dimensionalidad agresiva.

#### **4.4 Decisiones conscientes de selección y extracción**

A partir del análisis exploratorio realizado, se tomaron las siguientes decisiones metodológicas conscientes:

- **No aplicar reducción de dimensionalidad sobre los embeddings**  
El análisis PCA mostró que las primeras componentes explican una fracción moderada de la varianza total, lo cual es característico de representaciones semánticas distribuidas. Reducir dimensionalidad podría eliminar información semántica relevante para la recuperación.

- **Preservar el espacio semántico completo para retrieval**  
Mantener los embeddings completos permite conservar relaciones semánticas finas entre pasajes bíblicos, fundamentales para sistemas RAG basados en similitud coseno.

- **Usar PCA únicamente con fines diagnósticos**  
El PCA se utilizó como herramienta exploratoria para verificar la estabilidad y diversidad del espacio vectorial, no como mecanismo de compresión o selección de características.

- **Separar claramente análisis exploratorio de decisiones productivas**  
Esta distinción garantiza reproducibilidad, trazabilidad y alineación con la fase de Data Preparation definida por CRISP-ML(Q).

Estas decisiones priorizan la fidelidad semántica del corpus y sientan una base sólida para las fases posteriores de modelado y evaluación.

---

> *“Porque Dios no es Dios de confusión, sino de paz.”*
>
> — 1 Corintios 14:33

---

# **Parte 5 – Dataset final para modelado (Listo para RAG)**

En esta sección se consolida el resultado del proceso de **Ingeniería de Características**, describiendo el **dataset final** que queda preparado para las fases posteriores de Modeling y Evaluation dentro de la metodología CRISP-ML(Q).

A diferencia de enfoques tradicionales donde se produce un único dataset monolítico, en LUMINA el resultado de la fase de *Data Preparation* es un **conjunto coherente de artefactos complementarios**, cada uno con un propósito específico:

- Representación semántica (embeddings).
- Información estructural y contextual (metadatos).
- Trazabilidad y reproducibilidad (manifest).

El objetivo de esta parte es **verificar, describir y dejar explícito:**

- qué datos existen al finalizar el Avance 2,
- qué contiene cada artefacto,
- y cómo se utilizarán en un sistema RAG.


#### **5.1 Inventario de artefactos finales**

Al finalizar el Avance 2, el proyecto LUMINA cuenta con los siguientes artefactos persistentes:

- **Dataset tabular de chunks (.parquet)**  
Contiene texto, metadatos canónicos y variables categóricas codificadas.

- **Embeddings semánticos (.npz)**  
Representaciones vectoriales densas de cada chunk, generadas por traducción.

- **Manifest de embeddings (.json)**  
Metadatos de trazabilidad del proceso de extracción semántica.

Estos artefactos constituyen el **dataset final para modelado**, aunque se almacenen en formatos distintos por razones de eficiencia y separación de responsabilidades.


#### **5.2 Verificación básica del dataset final**

Antes de avanzar a modelado, es buena práctica realizar una verificación mínima de consistencia para confirmar que:

- Los artefactos existen.
- Las dimensiones coinciden.
- Los identificadores están alineados.

Esta verificación **no evalúa desempeño**, solo integridad y coherencia.

In [ ]:
# ============================================================
# Parte 5.2 — Verificación básica del dataset final
# ============================================================

# --- Localizamos los archivos más recientes ---
parquet_files = sorted(
    [f for f in os.listdir(ARTIFACTS_DIR) if f.endswith(".parquet")]
)
embedding_files = sorted(
    [f for f in os.listdir(EMB_DIR) if f.endswith(".npz")]
)
manifest_files = sorted(
    [f for f in os.listdir(EMB_DIR) if f.endswith(".json")]
)

print("Último dataset tabular (.parquet):", parquet_files[-1])
print("Último archivo de embeddings (.npz):", embedding_files[-1])
print("Último manifest (.json):", manifest_files[-1])

# --- Carga de artefactos ---
# allow_pickle=True es necesario porque chunk_id es un array de strings (dtype=object)
df_final = pd.read_parquet(os.path.join(ARTIFACTS_DIR, parquet_files[-1]))
emb_data = np.load(
    os.path.join(EMB_DIR, embedding_files[-1]),
    allow_pickle=True
)
with open(os.path.join(EMB_DIR, manifest_files[-1]), "r", encoding="utf-8") as f:
    manifest = json.load(f)

# --- Verificaciones clave ---
print("\nShape dataset tabular:", df_final.shape)
print("Número de chunks en tabular:", df_final.shape[0])

print("\nShape embeddings RV1909:", emb_data["emb_rv1909"].shape)
print("Shape embeddings VBL:", emb_data["emb_vbl"].shape)

print("\nChunks según manifest:", manifest["n_chunks"])

# Alineación por chunk_id
ids_tabular = set(df_final["chunk_id"])
ids_embeddings = set(emb_data["chunk_id"])

print("\n¿IDs alineados?:", ids_tabular == ids_embeddings)
print("Diferencia (si existe):", len(ids_tabular.symmetric_difference(ids_embeddings)))

Último dataset tabular (.parquet): lumina_chunks_tabular_20260207_084138.parquet
Último archivo de embeddings (.npz): lumina_embeddings_20260207_084138.npz
Último manifest (.json): embedding_manifest_20260207_084138.json

Shape dataset tabular: (10058, 29)
Número de chunks en tabular: 10058

Shape embeddings RV1909: (10058, 384)
Shape embeddings VBL: (10058, 384)

Chunks según manifest: 10058

¿IDs alineados?: True
Diferencia (si existe): 0


#### Interpretación de resultados (5.2)

En esta parte se verificó la consistencia e integridad del dataset final generado durante la Ingeniería de Características. Los artefactos resultantes (dataset tabular, embeddings semánticos y manifest de trazabilidad) presentan una alineación perfecta en número de instancias, dimensionalidad y referencias internas.

Esta verificación confirma que el proceso de preparación de datos fue ejecutado correctamente y que el conjunto final está listo para ser utilizado en fases posteriores de modelado y evaluación dentro de un sistema RAG, cumpliendo con los principios de reproducibilidad, modularidad y control establecidos por la metodología CRISP-ML(Q).

---

> *“Endereza tus sendas, y todos tus caminos serán rectos.”*
>
> — Proverbios 4:26

---

# **Conclusiones finales — Avance 2: Ingeniería de Características**

El presente Avance 2 del proyecto **LUMINA – La Biblia Viva con IA** abordó de manera integral la fase de **Ingeniería de Características**, transformando un corpus bíblico estructurado en representaciones adecuadas para aprendizaje automático y recuperación semántica.

A partir de un preprocesamiento mínimo y consciente, se construyeron características estructurales y numéricas que permiten mitigar sesgos canónicos y preservar trazabilidad textual. Posteriormente, el corpus fue segmentado en unidades semánticas coherentes mediante un enfoque de **chunking adaptativo**, alineado con la diversidad de géneros bíblicos.

La extracción de **embeddings semánticos densos** permitió representar el significado profundo de cada pasaje en un espacio vectorial continuo, constituyendo el núcleo informativo del sistema RAG. Estas representaciones se complementaron con metadatos codificados y fueron persistidas en artefactos reproducibles y verificables.

Finalmente, se validó la consistencia del dataset final, confirmando que el proyecto se encuentra preparado para avanzar a las fases de **Modeling y Evaluation** según la metodología CRISP-ML(Q). Las decisiones tomadas priorizan fidelidad semántica, estabilidad técnica y extensibilidad futura del sistema.

---

> *“Porque el Señor da la sabiduría, y de su boca viene el conocimiento y la inteligencia.”*
>
> — Proverbios 2:6

---